crates/store/README.md

```diff
@@ -0,0 +1,829 @@
+
+  2. **`ReadLayer`**: This trait extends `Layer` and provides read-only access
+     to the data store. It defines methods for checking if a key exists
+     (`has()`), retrieving data (`get()`), and creating iterators (`iter()`).
+
+  3. **`WriteLayer`**: This trait further extends `ReadLayer` to provide write
+     capabilities. It includes methods for putting data (`put()`), deleting data
+     (`delete()`), applying transactions (`apply()`), and committing changes
+     (`commit()`).
+
+These traits form a hierarchy that allows layers to build upon each other,
+adding or modifying functionality at each level.
+
+#### The store as a base layer
+
+The `Store` struct serves as the foundational layer in this architecture. It
+implements both `ReadLayer` and `WriteLayer`, providing direct access to the
+underlying database. The `Store` is typically backed by a concrete database
+implementation, such as [RocksDB][].
+
+#### Specialised layers
+
+The library includes several specialised layers that provide additional
+functionality:
+
+  1. **Temporal layer**
+  
+     The `Temporal` layer is a key component of the library's transactional
+     capabilities. It acts as a buffer between the application and the
+     underlying store, accumulating changes without immediately applying them to
+     the base layer.
+
+     This allows for complex operations to be built up over time and then
+     committed atomically.
+  
+     Key features of the `Temporal` layer include:
+     
+       - **Shadowing reads**: When reading data, it first checks its internal
+         shadow store before falling back to the base layer.
+     
+       - **Buffering writes**: All write operations are stored in the shadow
+         store rather than being immediately applied.
+     
+       - **Atomic commit**: When the `commit()` method is called, all
+         accumulated changes are applied to the base layer in a single
+         operation.
+  
+     The `Temporal` layer is particularly useful for scenarios where you need to
+     make multiple related changes and ensure they are all applied together, or
+     not at all.
+  
+  2. **Read-only layer**
+  
+     The `ReadOnly` layer provides a way to restrict access to read-only
+     operations. This is useful for ensuring that certain parts of your code
+     cannot accidentally modify the store, enhancing safety and preventing
+     unintended side effects.
+  
+     The `ReadOnly` layer implements `ReadLayer` but not `WriteLayer`,
+     effectively preventing any write operations from being performed through
+     it.
+
+#### Layer composition
+
+One of the most powerful aspects of the layer system is the ability to compose
+layers. Layers can be stacked on top of each other, each adding or modifying
+functionality. For example, you might create a `Temporal` layer on top of a
+`ReadOnly` layer, allowing for transactional-style operations without the
+ability to actually modify the underlying store.
```

Comment:
I feel a little disappointed, as I felt that was a cool scenario 😄 

There are situations where it can be useful to do this - for instance, dry runs of a sequence of events.

However, I think I got carried away, and didn't notice this isn't actually possible - and my example is perhaps not all that relevant to the target audience, I dunno.

Reply:
> dry runs of a sequence of events.

I can get behind that, we can patch it quite simply.

Especially since the embeddable model will expose these pieces to consumers, who can then choose to compose them how they want.

> and my example is perhaps not all that relevant to the target audience

it's spot on

> didn't notice this isn't actually possible

it's fine, let's continue with documenting "as-is", we'll revisit this piece